atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java

```diff
@@ -390,7 +404,66 @@ public RowColumnRangeIterator getRowsColumnRange(TableReference tableRef,
                         LoggingArgs.durationMillis(stopwatch)));
     }
 
-    private  <T> T maybeLog(Supplier<T> action, BiConsumer<LoggingFunction, Stopwatch> logger) {
+    private RowColumnRangeIterator wrapRowColumnRangeIteratorWithLogger(
+            Supplier<RowColumnRangeIterator> rowColumnRangeIteratorSupplier,
+            BiConsumer<LoggingFunction, Stopwatch> logger) {
+        RowColumnRangeIterator rowColumnRangeIterator = rowColumnRangeIteratorSupplier.get();
+        return new RowColumnRangeIterator() {
+            @Override
+            public Entry<Cell, Value> next() {
+                return maybeLog(rowColumnRangeIterator::next, logger);
+            }
+
+            @Override
+            public boolean hasNext() {
+                return rowColumnRangeIterator.hasNext();
```

Comment:
do any of these `hasNext` impls actually load the next batch?

Reply:
They can. The backing store for at least some of these iterators is the AbstractIterator, which has:

```The initial invocation of hasNext() or next() calls this method, as does the first invocation of hasNext or next following each successful call to next. Once the implementation either invokes endOfData or throws an exception, computeNext is guaranteed to never be called again.```

I'll also wrap them.